Integrate themes

ppizarror opened the pull request:
As discussed in #162 this PR integrate themes for pygame.
This makes a lot of changes, from the Menu constructor to widgets. Also examples were updated.

Added ".copy()" feature to Themes, because each theme object must be different for each menus.

ppizarror commented:
Mmm I thought a menu could change the theme, and it would affect the rest of the menus. But as far as I can see the theme is read only. So copying can be optional if user wants to create new themes based on others...

ppizarror commented:
> Really good job!!
> I tested the examples, they are perfectly functional.
> 
> What is the reason for theme object has to be different for each menu? It is not really user-friendly, but if there is no choice... no choice :)
> 
> The `menu_opacity`option could be also removed from the menu and provided as fourth element of the `Theme.background_color` tuple (RGBA tuple)

Removed opacity from Menu. Also removed that .copy() thing. Also added support for alpha channel to all colors on the menu (not only background). Theme will extend opacity for simple colors

ppizarror commented:
Finally for themes: Different menubar themes

``MENUBAR_MODE_ADAPTATIVE`
![image](https://user-images.githubusercontent.com/12925256/79173883-ebb92f00-7dc6-11ea-96fc-b39b5d3a0107.png)

`MENUBAR_MODE_SIMPLE`
![image](https://user-images.githubusercontent.com/12925256/79173937-14412900-7dc7-11ea-9550-16d3d79c0a83.png)

`MENUBAR_MODE_TITLE_ONLY`
![image](https://user-images.githubusercontent.com/12925256/79173953-29b65300-7dc7-11ea-891d-fa65d92bb52a.png)

`MENUBAR_MODE_TITLE_ONLY_DIAGONAL`
![image](https://user-images.githubusercontent.com/12925256/79173970-38046f00-7dc7-11ea-99e8-6dd3d7f996ee.png)

`MENUBAR_MODE_NONE`
![image](https://user-images.githubusercontent.com/12925256/79173988-4eaac600-7dc7-11ea-8074-635ac2285cc2.png)

`MENUBAR_MODE_UNDERLINE`
![image](https://user-images.githubusercontent.com/12925256/79174481-d513d780-7dc8-11ea-8750-708bfed6f47f.png)

`MENUBAR_MODE_UNDERLINE_TITLE`
![image](https://user-images.githubusercontent.com/12925256/79174493-e65ce400-7dc8-11ea-834b-8af238b6fa61.png)

anxuae commented:
I found 3 optimizations to discuss:

- the cursor color and text selection color are no more customizable (not in theme and not in the `add_text_input` arguments), thus no cursor and no selection on `THEME_BLACK`, maybe add theme to the theme? `widget_selection_color`, `widget_cursor_color`

- rename `widget_selection_effect` to `selection_effect` this option is related to the menu to render the selected widget, this is not consumed by the widget itself like all other `widget_...` parameters

- I update the doc of callbacks. For selector I think that we could simplify the passed arguments: by removing the index or the current text. By the way `Selector.get_value()` and  `Selector.set_value()` should remain consistent (set/get the index or set/get the text)

anxuae commented:
> Need some help for docs @anxuae thanks!!!

That's all for me

anxuae commented:
Maybe the `MENUBAR_STYLE_...` and `SOUND_TYPE_...` could be moved in the `locals` module? It is easier to find for user

ppizarror commented:
> Maybe the `MENUBAR_STYLE_...` and `SOUND_TYPE_...` could be moved in the `locals` module? It is easier to find for user

Fonts too?

ppizarror commented:
> Maybe the `MENUBAR_STYLE_...` and `SOUND_TYPE_...` could be moved in the `locals` module? It is easier to find for user

I think locals as a general file to store some constants through the application. Maybe we should keep sounds in sound, menubar styles in the menubar file, etc... pygame does something similar to this. For example

````
pygame.controls.LOCAL_CONTROLS..
pygame.image.LOCAL_IMAGES...
```

anxuae commented:
> > Maybe the `MENUBAR_STYLE_...` and `SOUND_TYPE_...` could be moved in the `locals` module? It is easier to find for user
> 
> I think locals as a general file to store some constants through the application. Maybe we should keep sounds in sound, menubar styles in the menubar file, etc... pygame does something similar to this. For example
> 
> ```
> pygame.controls.LOCAL_CONTROLS..
> pygame.image.LOCAL_IMAGES...
> ```
> ```

Ok I agree, following `pygame` concept is better

ppizarror commented:
Added widget background color, now in the near future both "background_color" and "widget_background_color" could be images. If color, just surface.blit, if "BaseImage", draw the image with repeatx, repeaty, stretch, etc...

ppizarror commented:
Working on image support:

![image](https://user-images.githubusercontent.com/12925256/79287548-1fad5680-7e92-11ea-81bd-4f099e720fd8.png)

ppizarror commented:
Added the following image drawing modes: `IMAGE_MODE_CENTER`, `IMAGE_MODE_FILL` (default), `IMAGE_MODE_REPEAT_X`, `IMAGE_MODE_REPEAT_XY`, `IMAGE_MODE_REPEAT_Y`, `IMAGE_MODE_SIMPLE`

ppizarror commented:
See updated example `Menu in textures and columns`

![image](https://user-images.githubusercontent.com/12925256/79292744-da902100-7e9f-11ea-9985-86d80d15d6d4.png)

ppizarror commented:
By now widgets only support FILL method because the surface is common through all the widgets. Background is quite easy because the surface is unique so clipping is enabled.

anxuae commented:
Good improvement!
Personally I don't like repeated patterns, it makes me think to my old Windows 3.11 wallpapers, yes it make me old ;-). But probably some people will want them.

anxuae commented on pygameMenu/baseimage.py:
```diff
@@ -30,21 +30,33 @@
 
 # Available images
-__actualpath = str(_path.abspath(_path.dirname(__file__))).replace('\\', '/')
+__actualpath = str(path.abspath(path.dirname(__file__))).replace('\\', '/')
```
- `.replace('\\', '/')` should not be necessary, the `os.path` library automatically use the correct path separator depending on the OS.
- The `str` is useless.
- The `abspath` should be applied before `dirname` as the content of the var `__file__` may vary depending on the OS
- don't concatenate path yourself, use the `join` method which acts depending on the OS

Use:
```
import os.path as osp
FONTDIR = osp.join(osp.dirname(osp.abspath(__file__)), '{0}', 'resources', 'images', '{1}')
```

ppizarror replied:
nice ⭐

anxuae commented on pygameMenu/baseimage.py:
```diff
@@ -30,21 +30,33 @@
+IMAGE_EXAMPLE_PYGAME_MENU = __fontdir.format(__actualpath, 'pygame_menu.png')
+
+# Drawing modes
+IMAGE_MODE_CENTER = 0
```
The integer values of  `IMAGE_MODE...`are not unique, same values used for `MENUBAR_STYLE...`

ppizarror replied:
Fixed!

anxuae commented on pygameMenu/sound.py:
```diff
@@ -30,22 +30,22 @@
+    AUDIO_ALLOW_FREQUENCY_CHANGE = False
 
-__actualpath = str(_path.abspath(_path.dirname(__file__))).replace('\\', '/')
+__actualpath = str(path.abspath(path.dirname(__file__))).replace('\\', '/')
```
Same remark as in baseimage.py

ppizarror replied:
Fixed!

anxuae commented on pygameMenu/sound.py:
```diff
@@ -69,7 +69,7 @@
 
 # Stores global reference that marks sounds as initialized
-SOUND_INITIALIZED = [False]
+_SOUND_INITIALIZED = [False]
```
It could be a class attribute of the `Sound` class. Global variables are generally not intended to be private.

```python
class Sound(object):

    sound_initialized = [0]

    def __init__(self, ....
```

then access using `Sound.sound_initialized`

ppizarror replied:
> Good improvement!
> Personally I don't like repeated patterns, it makes me think to my old Windows 3.11 wallpapers, yes it make me old ;-). But probably some people will want them.

hahaha I started with w98 playing captain claw and quake

ppizarror replied:
> It could be a class attribute of the `Sound` class. Global variables are generally not intended to be private.
> 
> ```python
> class Sound(object):
> 
>     sound_initialized = [0]
> 
>     def __init__(self, ....
> ```
> 
> then access using `Sound.sound_initialized`

Thay may work on Java and python 3, but python 2 throws an exception

anxuae commented on pygameMenu/widgets/widget/label.py:
```diff
@@ -55,6 +55,8 @@ def _apply_font(self):
     # noinspection PyMissingOrEmptyDocstring
     def draw(self, surface):
         self._render()
+        if self._label != '':
```
Why this condition `if self._label != '':` ? Even if there is no text it should be possible to se a background color/image?

ppizarror replied:
Yes, because the minimal surface width must be 1

anxuae commented:
@ppizarror I can not set anymore menu size to take the whole surface

<img width="795" alt="bug_margin" src="https://user-images.githubusercontent.com/25277953/79351903-83d72580-7f39-11ea-856c-6209a15a6948.png">

ppizarror commented:
> @ppizarror I can not set anymore menu size to take the whole surface. There is an irreducible margin.
> 
> <img alt="bug_margin" width="795" src="https://user-images.githubusercontent.com/25277953/79351903-83d72580-7f39-11ea-856c-6209a15a6948.png">

Fixed. The responsible for such bug was `self._scroll.set_position(self._pos_x, self._pos_y + menubar_height + 5)` It added extra 5 px margin. I refactored all margins, so the positions are clearer now

ppizarror commented:
I will merge this, this PR is way too large, as usual.
thanks!!
